pkg/issuer/acme/http/http.go

```diff
@@ -70,28 +75,43 @@ type reachabilityTest func(ctx context.Context, url *url.URL, key string) error
 // NewSolver returns a new ACME HTTP01 solver for the given Issuer and client.
 // TODO: refactor this to have fewer args
 func NewSolver(ctx *controller.Context) *Solver {
+	dynamicInformer := ctx.DynamicSharedInformerFactory.ForResource(istio.VirtualServiceGvr())
 	return &Solver{
-		Context:          ctx,
-		podLister:        ctx.KubeSharedInformerFactory.Core().V1().Pods().Lister(),
-		serviceLister:    ctx.KubeSharedInformerFactory.Core().V1().Services().Lister(),
-		ingressLister:    ctx.KubeSharedInformerFactory.Networking().V1beta1().Ingresses().Lister(),
-		testReachability: testReachability,
-		requiredPasses:   5,
+		Context:              ctx,
+		podLister:            ctx.KubeSharedInformerFactory.Core().V1().Pods().Lister(),
+		serviceLister:        ctx.KubeSharedInformerFactory.Core().V1().Services().Lister(),
+		ingressLister:        ctx.KubeSharedInformerFactory.Networking().V1beta1().Ingresses().Lister(),
+		virtualServiceLister: dynamiclister.New(dynamicInformer.Informer().GetIndexer(), istio.VirtualServiceGvr()),
+		testReachability:     testReachability,
+		requiredPasses:       5,
 	}
 }
 
 func http01LogCtx(ctx context.Context) context.Context {
 	return logf.NewContext(ctx, nil, "http01")
 }
 
-func httpDomainCfgForChallenge(ch *cmacme.Challenge) (*cmacme.ACMEChallengeSolverHTTP01Ingress, error) {
+func httpIngressForChallenge(ch *cmacme.Challenge) (*cmacme.ACMEChallengeSolverHTTP01Ingress, error) {
 	if ch.Spec.Solver.HTTP01 == nil || ch.Spec.Solver.HTTP01.Ingress == nil {
-		return nil, fmt.Errorf("challenge's 'solver' field is specified but no HTTP01 ingress config provided. " +
+		return nil, errors.New("challenge's 'solver' field is specified but no HTTP01 ingress config provided. " +
 			"Ensure solvers[].http01.ingress is specified on your issuer resource")
 	}
 	return ch.Spec.Solver.HTTP01.Ingress, nil
 }
 
+func serviceTypeForChallenge(ch *cmacme.Challenge) (corev1.ServiceType, error) {
+	if ch.Spec.Solver.HTTP01 != nil {
+		if ch.Spec.Solver.HTTP01.Ingress != nil {
+			return ch.Spec.Solver.HTTP01.Ingress.ServiceType, nil
+		}
+		if ch.Spec.Solver.HTTP01.Istio != nil {
+			return corev1.ServiceTypeClusterIP, nil
+		}
+	}
+
+	return "", errors.New("could not determine service type for challenge")
+}
```

Comment:
I was curious about why the user would want to choose the `Ingress.ServiceType` but not for `Istio`? Please find out the motivation for adding `Ingress.ServiceType` to the API and then add a TODO comment or an explanatory comment, saying why this isn't needed for the Istio VirtualService.

Reply:
We probably should allow other Service Types here too? This is something that was implemented by @khernyo. I'll add a field to the Istio object to also allow adding a ServiceType option!

Reply:
I think this is the commit which introduced the `Ingress.ServiceType` field (as `HTTP01.SolverServiceType` at that time): https://github.com/jetstack/cert-manager/commit/d261e1f3f1afa069f381ba516936596193737267

For Istio, `ClusterIP` should be enough because the ACME request enters the cluster through an envoy proxy which should have an accompanying service with the proper service type. This envoy proxy forwards the request to the solver pod. As the envoy proxy is running inside the cluster, the solver service only needs `ClusterIP` service type.

(If the service for the envoy proxy has an incorrect service type, that's a problem, but it can't be fixed by cert-manager)

I can't think of a scenario where a `ClusterIP` service type wouldn't be enough. Although making it configurable shouldn't hurt much, it does increase code and configuration complexity. I would suggest to avoid making it configurable now, and add it in a future cert-manager version if it really is necessary.